### ISSUE TITLE
Stack overflow in conn_recv_handshake_pkt

### ISSUE THREAD

**Issue by Freddrickk** [ngtcp2/ngtcp2#274]
Hello,

I'm opening this issue to report a stack overflow in conn_recv_handshake_pkt. It occurs on a QUIC server when parsing a packet of type `NGTCP2_PKT_INITIAL` during the handshake phase. The buggy code path can only be reached when when connection object's field `conn->local.settings.token.len` is set to zero. The client must send a packet with a header containing a very large token field. Then, the number of bytes read from the header by `ngtcp2_pkt_decode_hd_long` and assigned to `nread` will be very large. When `decrypt_hp` is called, the `nread` value is larger than the stack buffer `plain_hdpkt` which has a static size of 1500 bytes:

```cpp
static ngtcp2_ssize conn_recv_handshake_pkt(ngtcp2_conn *conn,
                                            const ngtcp2_path *path,
                                            const uint8_t *pkt, size_t pktlen,
                                            ngtcp2_tstamp pkt_ts,
                                            ngtcp2_tstamp ts) {
[...]
uint8_t plain_hdpkt[1500];
[...]
nread = ngtcp2_pkt_decode_hd_long(&hd, pkt, pktlen);
[...]
nwrite = decrypt_hp(&hd, plain_hdpkt, sizeof(plain_hdpkt), hp, pkt, pktlen,
                    (size_t)nread, ckm, hp_ctx, hp_mask);
[...]
```

`decrypt_hp` then copies the content of the packet into this buffer using `nread` as the size to copy leading to a stack overflow.

```cpp
static ngtcp2_ssize decrypt_hp(ngtcp2_pkt_hd *hd, uint8_t *dest, size_t destlen,
                               const ngtcp2_crypto_cipher *hp,
                               const uint8_t *pkt, size_t pktlen,
                               size_t pkt_num_offset, ngtcp2_crypto_km *ckm,
                               const ngtcp2_crypto_cipher_ctx *hp_ctx,
                               ngtcp2_hp_mask hp_mask) {
  size_t sample_offset;
  uint8_t *p = dest;
  uint8_t mask[NGTCP2_HP_MASKLEN];
  size_t i;
  int rv;

  assert(hp_mask);
  assert(ckm);
  assert(destlen >= pkt_num_offset + 4); // asserts are disabled on non-debug builds

  if (pkt_num_offset + 4 + NGTCP2_HP_SAMPLELEN > pktlen) {
    return NGTCP2_ERR_PROTO;
  }

  p = ngtcp2_cpymem(p, pkt, pkt_num_offset); // The overflow occurs here
  [...]
```

This issue may also affect `conn_recv_pkt` which also uses the static-sized `uint8_t plain_hdpkt[1500]` buffer . It would be best to apply the same patch in this function too.

Also, since QUIC drafts do not specify a maximum size for a retry token so it's also possible that a server sets a very long retry token which would also lead to a stack buffer overflow.

Regards,

**Comment by tatsuhiro-t** [comment]
Thank you for reporting this issue.  It is definitely a bug of early false assumption.  Will fix it soon.

**Comment by tatsuhiro-t** [comment]
Fix committed via 0216f1153b1c6803b915b255f05a41a46f1bf8f0
